cubedash/index/postgis/_api.py

```diff
@@ -0,0 +1,977 @@
+                    ).label("month"),
+                    func.count(),
+                )
+                .select_from(ODC_DATASET)
+                .where(
+                    and_(
+                        ODC_DATASET.product_ref == product.id,
+                        column("updated") > only_those_newer_than,
+                    )
+                )
+                .group_by("month")
+                .order_by("month")
+            )
+
+    def outdated_years(self, product_id: int):
+        updated_months = aliased(TimeOverview, name="updated_months")
+        years = aliased(TimeOverview, name="years_needing_update")
+
+        with self.index._active_connection() as conn:
+            return conn.execute(
+                # Select years
+                select(years.start_day)
+                .where(years.period_type == "year")
+                .where(
+                    years.product_ref == product_id,
+                )
+                # Where there exist months that are more newly created.
+                .where(
+                    exists(
+                        select(updated_months.start_day)
+                        .where(updated_months.period_type == "month")
+                        .where(
+                            func.extract("year", updated_months.start_day)
+                            == func.extract("year", years.start_day)
+                        )
+                        .where(
+                            updated_months.product_ref == product_id,
+                        )
+                        .where(updated_months.generation_time > years.generation_time)
+                    )
+                )
+            )
+
+    def product_ds_count_per_period(self):
+        with self.index._active_connection() as conn:
+            return conn.execute(
+                select(
+                    ProductSpatial.name,
+                    TimeOverview.start_day,
+                    TimeOverview.period_type,
+                    TimeOverview.dataset_count,
+                )
+                .select_from(TimeOverview)
+                .join(ProductSpatial)
+                .where(TimeOverview.product_ref == ProductSpatial.id)
+                .order_by(
+                    ProductSpatial.name,
+                    TimeOverview.start_day,
+                    TimeOverview.period_type,
+                )
+            )
+
+    def upsert_product_record(self, product_name: str, fields):
+        # Dear future reader. This section used to use an 'UPSERT' statement (as in,
+        # insert, on_conflict...) and while this works, it triggers the sequence
+        # `product_id_seq` to increment as part of the check for insertion. This
+        # is bad because there's only 32 k values in the sequence and we have run out
+        # a couple of times! So, It appears that this update-else-insert must be done
+        # in two transactions...
+        with self.index._active_connection() as conn:
```

Comment:
Doing in two SQL statements makes sense, but could (and probably should) still be done in a single transaction:

Just use `self.index.transaction()` instead of `self.index._active_connection()`